v1 (#8)
NHS-NSS-transforming-publications/git-guide: 01-intro.Rmd

# Introduction & Setup {-}

Who for?
What
Where
When

> Why

Author: This was actually meant as a reminder for me to come and write something meaningful here - sorted now :)

## Version Control, Git, and Remotes {-}

#### Version Control {-}

A version control system, or VCS, tracks the history of changes as people and teams collaborate on projects together. As the project evolves, teams can run tests, fix bugs, and contribute new code with the confidence that any version can be recovered at any time. Developers can review project history to find out:

* Which changes were made?
* Who made the changes?
* When were the changes made?
* Why were changes needed?

Software teams that do not use any form of version control often run into problems; changes not available to users or creation of incompatible changes. If you're a developer who has never used version control you may have added versions to your files, perhaps with suffixes like "final" or "latest" and then had to later deal with a new final version. Perhaps you've commented out code blocks because you want to disable certain functionality without deleting the code, fearing that there may be a use for it later. Version control is a way out of these problems.

#### Git {-}

> Git is a version control system (VCS). We recommend that git is used along with a hosting repository (a remote), such as GitHub (or Gitea, GitLab, etc.) and that the [GitHub Flow] is followed. However, git can be used locally while working on a project. The steps outlined in [Quick Start Git] are for this purpose and should only be followed for a personal project that is **local or on your personal network folder**. For collaborative work, see the getting started section for [Start GitHub].

Author: I had this is as an introduction/general overview for awareness - the next section, GitHub workflow, references remotes so wanted people to know what it was talking about. I've adjusted the wording to reflect this.

Systems like Git allow us to move work between any two repositories. In practice, though, it’s easiest to use one copy as a central hub, and to keep it on the web rather than on someone’s laptop. This is where remotes come in. Most programmers use hosting services like [GitHub](https://github.com/), [BitBucket](https://bitbucket.org/) or [GitLab](https://gitlab.com/) to hold those master copies. We currently use **GitHub.**

GitHub is a git hosting repository that provides users with tools to ship better code through command line features, issues (threaded discussions), pull requests, and code review. GitHub builds collaboration directly into the development process. Work is organised into repositories, where users can outline requirements or direction and set expectations for team members. Then, using the [GitHub Flow], developers simply create a branch to work on updates, commit changes to save them, open a pull request to propose and discuss changes, and merge pull requests once everyone is on the same page. See [GitHub Workflow] for more details.

## Workflows {-}

### GitHub Workflow {-}

> ```{r fig.align='center', echo=FALSE, include=identical(knitr:::pandoc_to(), 'html')}

Author: https://guides.github.com/pdfs/githubflow-online.pdf - the PDF version of the GitHub workflow guide.  I've added as a caption.

knitr::include_graphics('images/githubflow.png', dpi = NA)
```
&nbsp;

This is our recommended workflow. The workflow is branch-based and involves keeping the master branch clean and 'production ready'/fully-functional at all times. 

1. **Create a branch** - topic/per-person branches created from the master allow teams to contribute to many parallel efforts. Short-lived topic branches, in particular, keep teams focussed and results in quick ships.
2. **Add commits** - these are snapshots of development efforts within a branch and create safe, revertible points in the project's history.
3. **Open a pull request** - pull requests publicise the project's ongoing efforts and set the tone for a transparent development process.

> 4. **Discuss and review code** - teams participate in code reviews by commenting, testing, and reviewing open pull requests. Code review is at the centre of an open and participatory culture. Pull requests should be approved by at least one collaborator - we recommend you assign a specific reviewer(s) to check the work. In order to ensure the master branch remains 'production ready' we also recommend that the master branch is protected and any changes are tested by the reviewer before approving.

Reviewer: ```suggestion
4. **Discuss and review code** - teams participate in code reviews by commenting, testing, and reviewing open pull requests. Pull requests should be approved by at least one collaborator - we recommend you assign a specific reviewer(s) to check the work. In order to ensure the master branch remains 'production ready' we also recommend that the master branch is protected and any changes are tested by the reviewer before approving.
```

> 5. **Merge** - upon clicking merge, GitHub automatically performs the equivalent of a local `git merge` operation. GitHub also keeps the entire branch development history on the merged pull request.

Reviewer: I don't think equating this to the local command is particularly necessary here - it's unlikely anyone will ever have to use it!

> 6. **Deploy** - teams can choose the best release cycles or incorporate continuous integration tools and operate with the assurance that code on the deployment branch has gone through a robust workflow.

Reviewer: This is an optional steps, so I would make it clear teams don't have to actually do this.

## Setup {-}

### Git {-}

*These intructions should only need to be followed once or when setting up a new device.*

> 1. **Install Git** - Git is a free, open-source, software available from https://git-scm.com/. If you work within NHS National Services Scotland, request the software from IT through the [Service Portal](https://nhsnss.service-now.com/sp). Once authorised, this will allow you to download it from the Software Center on your machine. 

Reviewer: ```suggestion
1. **Install Git** - Git is a free, open-source, software which can be requested from the software from IT through the [Service Portal](https://nhsnss.service-now.com/sp). Once authorised, this will allow you to download it from the Software Center to your machine. 
```

This will install three applications on your computer, go to Start > Git (folder):

* **Git Bash** - this is the command line (Shell) interface for git. This will allow you to enter git commands, i.e. `git init`, and can be launched by right clicking in a directory/folder and selecting "Git Bash Here".
* **Git CMD** - this is a deprecated version, similar to the Bash interface, and shouldn't be used here.

> * **Git GUI** - this is git's version of a Graphical User Interface. This will perform the same functions as "Git Bash" by the user selecting options rather than having to type in the commands. This guide will outline using integrations on software like RStudio to utilise GUIs rather than this option.

Reviewer: ```suggestion
* **Git GUI** - this is git's version of a Graphical User Interface. This will perform the same functions as "Git Bash", but employs a point-and-click interface instead. This guide does not cover Git GUI and instead focuses on the integrated Git GUI in RStudio.
```

> 2. **Configure your details** - Git needs to know who you are, use the following commands to configure your username and email (*if you're a GitHub user, use the same email and username you're registered with on GitHub.*):

Reviewer: Might also be worth adding a note that this step is necessary for GitHub to recognise the account commits come from. Even when pushing with proper authentication, GitHub assigns commits to an account via the email address attached, so these need to match!

> * `git config --global user.email <NHS email address>`

Reviewer: The earlier version of this guide had some text stating the <> should be removed when adding your own details. Might be worth adding back in somewhere? Apologies if it has just been moved somewhere I haven't reached yet!

* `git config --global user.name <your name>`

At any time you can check what the current user details are using:

* `git config --global user.email`

### GitHub {-}

These instructions are for first-time users of GitHub and highlight the steps in setting up a GitHub account.

> 1. **Sign up** - In order to use GitHub, you'll need to have a personal GitHub account, you can set that up on the [GitHub website](http://github.com). If using your NHS email address ensure you **don't use the same password as your NHS email account**. 

Reviewer: People also shouldn't be using their LDAP passwords, especially if their username is the same as their LDAP login.

2. **Edit your details** - you will be able to set these items up as part of the sign up process. Otherwise, follow the links and edit them now.

* Go to your [Profile Settings](https://github.com/settings/profile) and add your details with a picture of yourself if you feel comfortable to do so.
* Go to your [Account Settings](https://github.com/settings/admin) and make sure your username is set to be your full name. This helps to identify users as generated GitHub handles tend to pretty obscure.

3. **Get added to Organisations** - this step is optional.

* If you're doing work within the Transforming Publishing team. Email one of the admins [here](https://github.com/orgs/NHS-NSS-transforming-publications/teams/transforming-publishing-team/members?utf8=%E2%9C%93&query=+role%3Amaintainer) to request being added to the organisation.
* If you work within Public Health & Intelligence (PHI), the [Health & Social Care Scotland](https://github.com/Health-SocialCare-Scotland) GitHub may be of interest. Repositories can be requested and set up for team working.

### RStudio {-}

> First [R](https://www.r-project.org/) and [RStudio](https://www.rstudio.com/) must be installed on your machine. Just as with git, if you work within NHS National Services Scotland these can be requested through the [Service Portal](https://nhsnss.service-now.com/sp). Once authorised, this will allow you to download it from the Software Center on your machine. You may also be able to use the server versions of RStudio and the project workflow will be the same.

Reviewer: ```suggestion
First [R](https://www.r-project.org/) and [RStudio](https://www.rstudio.com/) must be installed on your machine. Just as with git, these can be requested through the [Service Portal](https://nhsnss.service-now.com/sp). Once authorised, this will allow you to download it from the Software Center on your machine. You may also be able to use the server versions of RStudio and the project workflow will be the same.
```

1. **Once in RStudio, go to Tools in the menu bar > Global Options > Git/SVN**:

```{r fig.align='center', echo=FALSE, include=identical(knitr:::pandoc_to(), 'html')}
knitr::include_graphics('images/rstudio-globalopt-gitsvn.png', dpi = NA)
```
&nbsp;

> 2. **If "Git executable:" shows "*(none)*", click "Browse" and select the git executable installed on your system.** - On Windows, git.exe will likely be be in "Program Files", specifically try `C:/Program Files/Git/bin/git.exe`. That's it for attaching RStudio and Git. This will allow you to utilise the RStudio Git GUI tools inside R Projects. 

Reviewer: Occasionally, there is a filepath on the server, but it points to the wrong folder. I would highlight what the filepath should be for the server (/usr/bin/git) and then for the desktop just say the filepath should point to the .../bin/git.exe file in the folder git was installed to. This is quite dependent on the user and their OS so I wouldn't provide an example as that might confuse people where it is different. Might be worth pointing them towards somewhere they can get help if they can't find it, though.
